Document postgres event store (#1534)
sensu/sensu-docs: content/sensu-go/5.10/reference/datastore.md

## Scaling event storage

**LICENSED TIER**: Unlock enterprise-scale event storage in Sensu Go with a Sensu license. To activate your license, see the [getting started guide][13].

Sensu supports using an external PostgreSQL instance for event storage in place of etcd.
PostgreSQL can handle significantly higher volumes of Sensu events, letting you scale Sensu beyond etcd's 8GB limit.

Configured with a PostgreSQL event store, Sensu connects to PostgreSQL to store and retrieve event data in place of etcd, while etcd continues to store Sensu configuration data.
You can access event data stored in PostgreSQL using the same Sensu web UI, API, and sensuctl processes as etcd-stored events.

### Requirements

Sensu supports PostgreSQL 9.5 and later as well as [Amazon Relational Database Service][3] (Amazon RDS).
See the [PostgreSQL docs][14] to install and configure PostgreSQL.

### Configuration

At the time of enabling the PostgreSQL event store, event data cuts over from etcd to PostgreSQL, resulting in a loss of recent event history.
No restarts or Sensu backend configuration changes are required to enable the PostgreSQL event store.

> _WARNING: At this time, the PostgreSQL event store cannot be disabled once configured._

Reviewer: Should we mention that we plan to address this in the subsequent release?

Author: Usually we try to avoid that sort of info in the docs, but I think in this case, a friendly sentence to that effect would be helpful. Good callout!

After installing and configuring PostgreSQL, configure Sensu by creating a `PostgresConfig` resource. See the [specification](#specification) for more information.

{{< language-toggle >}}

{{< highlight yml >}}
type: PostgresConfig
api_version: store/v1
metadata:
  name: my-postgres
spec:
  dsn: "postgresql://user:secret@host:port/dbname"
  pool_size: 20
{{< /highlight >}}

{{< highlight json >}}
{
  "type": "PostgresConfig",
  "api_version": "store/v1",
  "metadata": {
    "name": "my-postgres"
  },
  "spec": {
    "dsn": "postgresql://user:secret@host:port/dbname",
    "pool_size": 20
  }
}
{{< /highlight >}}

{{< /language-toggle >}}

With the `PostgresConfig` resource definition saved to a file (for example: postgres.yml), use sensuctl, [configured as the admin user][1], to activate the PostgreSQL event store.

{{< highlight shell >}}
sensuctl create --file postgres.yml
{{< /highlight >}}

At this time, there is no supported method for viewing or deleting a `PostgresConfig` resource.
To update your Sensu PostgreSQL configuration, repeat the `sensuctl create` process shown above.

> If Sensu cannot connect to a configured PostgreSQL instance, you can expect to see errors in the [Sensu backend logs][2] at the `warn` log level.

Reviewer: Errors will be written at the `ERROR` log level, and status updates (like for the store being enabled, updated, or disabled) will be issued at `WARN` level.

---
title: "Datastore"
description: "Sensu stores the most recent event for each entity and check pair using an embedded etcd (default) or an external etcd instance. Sensu also supports using an external PostgreSQL instance for event storage in place of etcd (licensed-tier only). Read the reference to configure enterprise-scale event storage using PostgreSQL."
weight: 10
version: "5.10"
product: "Sensu Go"
menu:
  sensu-go-5.10:
    parent: reference
---

- [Event storage](#event-storage)
- [Scaling event storage](#scaling-event-storage) (licensed-tier only)
  - [Requirements](#requirements)
  - [Configuration](#configuration)
  - [Specification](#specification)

## Event storage

Sensu stores the most recent event for each entity and check pair using an embedded etcd (default) or an [external etcd][8] instance.
You can access event data using the [dashboard][9] events page, [`sensuctl event` commands][10], and the [events API][11].

> To store event data over time, we recommend integrating Sensu with a time series database, like [InfluxDB][12].

Reviewer 2: ```suggestion
For longer retention of event data, we recommend integrating Sensu with a time series database, like [InfluxDB][12], or a searchable index, like ElasticSearch or Splunk.
```

Author: Thanks, @cwjohnston! Can you elaborate on this suggestion?

Reviewer 2: In most cases where I've seen folks integrating Sensu with tsdb for extended event storage, the data stored in a tsdb tends to be a very lossy representation of the event. Typically this is the timestamp and a selection of attributes such as the check name, entity name and check result status. 

This is usually enough context for overlaying monitoring events on graphs which illustrate related metrics/telemetry, and in many cases this is probably a reasonable tradeoff, as I imagine adding a large number of labels/tags to each measurement isn't a recommended practice.

On the other hand, indexers like elasticsearch and splunk are typically considered a reasonable technology for recording complete event records. Right now this should be doable using our licensed feature for recording events to a flat file, using something like ElasticSearch filebeat or the Splunk log collector to store that data. In recent years these indexers have made strides to become better data sources for metrics/telemetry, making it possible that a team could choose to adopt one of these instead or or in addition to a timeseries database.

Author: This is great context, thanks!

## Scaling event storage

**LICENSED TIER**: Unlock enterprise-scale event storage in Sensu Go with a Sensu license. To activate your license, see the [getting started guide][13].

Sensu supports using an external PostgreSQL instance for event storage in place of etcd.
PostgreSQL can handle significantly higher volumes of Sensu events, letting you scale Sensu beyond etcd's 8GB limit.

> Configured with a PostgreSQL event store, Sensu connects to PostgreSQL to store and retrieve event data in place of etcd, while etcd continues to store Sensu configuration data.

Reviewer 2: ```suggestion
Configured with a PostgreSQL event store, Sensu connects to PostgreSQL to store and retrieve event data in place of etcd, while etcd continues to store Sensu entity and configuration data.
```

Author: Thanks, @cwjohnston! Can you elaborate on this suggestion?

Reviewer 2: I'm thinking about this as a way to suggest or clarify that etcd is still being used as a registry for things which aren't events and also aren't configuration, strictly speaking.

You can access event data stored in PostgreSQL using the same Sensu web UI, API, and sensuctl processes as etcd-stored events.

### Requirements

> Sensu supports PostgreSQL 9.5 and later as well as [Amazon Relational Database Service][3] (Amazon RDS).

Reviewer 2: ```suggestion
Sensu supports PostgreSQL 9.5 and later, including [Amazon Relational Database Service][3] (Amazon RDS) when configured with the PostgreSQL engine.
```
